src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -962,35 +984,57 @@ private static Exception GetLastIOError()
             return Interop.GetIOException(Interop.Sys.GetLastErrorInfo());
         }
 
-        private sealed class SerialStreamIORequest : TaskCompletionSource<int>
+        private abstract class SerialStreamIORequest : TaskCompletionSource<int>
         {
-            public Memory<byte> Buffer { get; private set; }
             public bool IsCompleted => Task.IsCompleted;
             private CancellationToken _cancellationToken;
+            private readonly CancellationTokenRegistration _cancellationTokenRegistration;
 
-            public SerialStreamIORequest(CancellationToken ct, Memory<byte> buffer)
+            protected SerialStreamIORequest(CancellationToken ct)
                 : base(TaskCreationOptions.RunContinuationsAsynchronously)
             {
                 _cancellationToken = ct;
-                ct.Register(s => ((TaskCompletionSource<int>)s).TrySetCanceled(), this);
-
-                Buffer = buffer;
-            }
-
-            internal void Complete()
-            {
-                Debug.Assert(Buffer.Length == 0);
-                TrySetResult(Buffer.Length);
+                _cancellationTokenRegistration = ct.Register(s => ((TaskCompletionSource<int>)s).TrySetCanceled(), this);
             }
 
             internal void Complete(int numBytes)
             {
                 TrySetResult(numBytes);
+                _cancellationTokenRegistration.Dispose();
             }
 
             internal void Complete(Exception exception)
             {
                 TrySetException(exception);
+                _cancellationTokenRegistration.Dispose();
+            }
+        }
+
+        private sealed class SerialStreamReadRequest : SerialStreamIORequest
+        {
+            public Memory<byte> Buffer { get; private set; }
```

Comment:
we had similar issue in `FileStream` and we were able to always store `ReadOnlyMemory` and cast it to `Memory` using `MemoryMarshal.AsMemory`

https://github.com/dotnet/runtime/blob/6fc8441311120c47c96b2605bf3a0d186f82454a/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs#L98

Since this project targets .NET Standard 2.0 and .NET 4.6.1:

https://github.com/dotnet/runtime/blob/6fc8441311120c47c96b2605bf3a0d186f82454a/src/libraries/System.IO.Ports/src/System.IO.Ports.csproj#L7

We can't do it here as this method is not available for these TFMs:

https://apisof.net/catalog/61409471-3239-9645-a842-a812d650f79f

Reply:
> Since this project targets .NET Standard 2.0 and .NET 4.6.1:

Is that why some of the builds break? The memory overloads for Stream don't exist in 4.6.1. Should they be ifdef-ed?

Reply:
> Should they be ifdef-ed?

Yes